### ISSUE TITLE
delay responses

### ISSUE THREAD

**Issue by rolandjitsu** [wheresrhys/fetch-mock#75]
Is there any way to delay a response for a certain amount of time?

**Comment by wheresrhys** [comment]
Just been added yesterday : ). You can now respond with a promise for a response config object. so e.g.
```javascript
const delay = new Promise((res, rej) => setTimeout(res, 500))
fetchMock.mock('http://thing.com', delay.then(404))
```

let me know if it doesn't work (you'll probably be one of the first to use it)

**Comment by rolandjitsu** [comment]
Well, `fetchMock.mock('http://thing.com', delay.then(404))` is not valid syntax. I did try:
```js
const delay = new Promise((resolve) => {
	setTimeout(() => resolve({$uri: '/delayed/1', delay: 500}), 2500);
});
fetchMock.mock('http://localhost/delayed/1', delay);
```
But that does not seem to work.

**Comment by rolandjitsu** [comment]
Oh, I was at `4.1.0`, let me update and try again.

**Comment by rolandjitsu** [comment]
It does work, thanks.

**Comment by devarsh** [comment]
Consider this code snippet 

```js
var fetchMock = require('fetch-mock')

const delay = () => new Promise((res, rej) => setTimeout(res, 3000))

fetchMock.mock(/(http:\/\/localhost:8081\/response)[/.-\w\d]*/,
  delay().then(() => ({
      Ack : true
    }
})))



fetch('http://localhost:8081/response').then(data => data.json()).then(data => console.log(data))
setTimeout(()=> {
  console.log("i am making req")
  fetch('http://localhost:8081/response').then(data => data.json()).then(data => console.log(data))
}, 5000)
```

Subsequent call is fetch are resolved immediately and not respecting the timeout set

**Comment by fizwidget** [comment]
Hey @wheresrhys, first of all thanks for writing fetch-mock, it's come in really handy :)

I've also run into a few problems when trying to delay responses though...

As @devarsh found, it doesn't seem like it's easy (or at all possible?) to delay responses to subsequent `fetch` calls. A related problem is that the delay begins counting down when the promise is created, _not_ when the call to `fetch` occurs. I think these problems are simply due to the semantics of promises - there's no way of "re-running" a promise, and the body of a promise (the call to `setTimeout`) is executed when the promise is created.

There might be a way around this that I'm missing, would love to hear if anyone has one. My gut feeling though is that delays should be more explicitly supported by the library, e.g. a `delay` configuration option. Or perhaps instead of directly supplying a promise, we could supply a function that returns a new promise when called?

**Comment by fizwidget** [comment]
Ah nevermind, I'm an idiot :P

The API already allows for supplying a function that returns a promise, which solves all the above-mentioned problems 👍

**Comment by devarsh** [comment]
@fizwidget Thanks for writing out, well i'm using the delay parameter, in which i'm passing a function which returns the promise, i'm not passing the promise to the delay parameter, so my concern is to run delay fn on each incoming request rather than resolving it at once.

Which will solve the subsequent fetch request.

**Comment by fizwidget** [comment]
Hey @devarsh, I think I understand what you mean. This small change should fix your problem:

```
fetchMock.mock(/(http:\/\/localhost:8081\/response)[/.-\w\d]*/,
  () => delay().then(() => ({
      Ack : true
    }
})))
```

Notice how the second argument to `fetchMock.mock` is now a function that creates a promise, rather than a promise itself :)

**Comment by devarsh** [comment]
oops my argument was passing the promise and not the fn, silly mistake on my part

**Comment by thommyhh** [comment]
Even this is closed (I stepped upon it via Google search), I'd like to share the gist I created form my solution:

https://gist.github.com/thommyhh/71fa3bc49d9b38f84162bccea20e3522

**Comment by manikantag** [comment]
Complete solution put in one comment:

```
const delay = (response, after=500) => () =>
  new Promise(resolve => setTimeout(resolve, delayTime)).then(() => response);

fetchMock.mock('http://example.com', delay(200, 3000)); // send '200' response after 3000 millis

**Comment by alfonsomunozpomer** [comment]
There’s now [an open enhancement to implement this declaratively](https://github.com/wheresrhys/fetch-mock/issues/456). Go and up-vote it!

**Comment by wheresrhys** [comment]
Implemented in 7.7.0.
